bn-apps/cordapp-updates-distribution/corda-updates-core/src/main/kotlin/net/corda/businessnetworks/cordaupdates/core/CordaMavenResolver.kt

```diff
@@ -0,0 +1,248 @@
+     * @param configProps will be passed as custom Maven Resolver session properties
+     *
+     * @return metadata for all of the versions (both existing and downloaded). isFromLocal field indicates whether a version
+     *      existed locally or not
+     */
+    fun downloadVersionRange(coordinatesWithRange : String,
+                             configProps : Map<String, Any> = mapOf()) : ArtifactMetadata {
+        val artifactMetadata = resolveVersionRange(coordinatesWithRange, configProps)
+
+        val aetherArtifact = DefaultArtifact(coordinatesWithRange)
+
+        // downloading the most recent version first
+        val versions = artifactMetadata.versions.asReversed().map {
+            val artifactVersion = aetherArtifact.setVersion(it.version)!!
+            downloadVersion(artifactVersion.toString(), configProps).versions.single()
+        }.asReversed()
+        return artifactMetadata.copy(versions = versions)
+    }
+
+    /**
+     * Downloads a specified version of a CorDapp if it's missing in the local repo.
+     *
+     * @param coordinates full maven coordinates, i.e. "net.corda:corda-finance:3.2".
+     * @param configProps will be passed as custom Maven Resolver session properties
+     *
+     * @return artifact metadata with a single version
+     *
+     * @throws [ResourceNotFoundException] if the the specified artifact has not been found in the remote repository
+     * @throws [ResourceTransferException] if the remote repository is unreachable
+     */
+    fun downloadVersion(coordinates : String,
+                        configProps : Map<String, Any> = mapOf()) : ArtifactMetadata {
+        val session = createRepositorySession(configProps)
+        val aetherArtifact = DefaultArtifact(coordinates)
+        val artifactRequest = ArtifactRequest(aetherArtifact, listOf(remoteRepository), null)
+        val resolutionResult : ArtifactResult
+        try {
+            resolutionResult = repositorySystem.resolveArtifact(session, artifactRequest)!!
+        } catch (ex : RepositoryException) {
+            when (ex.cause) {
+                is ArtifactNotFoundException -> throw ResourceNotFoundException(coordinates, ex)
+                is ArtifactTransferException -> throw ResourceTransferException(coordinates, ex)
+                else -> throw CordaMavenResolverException("Error while resolving the artifact $coordinates", ex)
+            }
+        }
+        return ArtifactMetadata(aetherArtifact.groupId,
+                aetherArtifact.artifactId,
+                aetherArtifact.classifier,
+                aetherArtifact.extension,
+                listOf(VersionMetadata(aetherArtifact.version, resolutionResult.repository is LocalRepository)))
+    }
+
+    private fun createRepositorySession(additionalConfigProperties : Map<String, Any>) : DefaultRepositorySystemSession {
+        val session = newSession()!!
```

Comment:
A better way to write this is by specifying the type explicitly (alt-enter, x, enter). If it's not nullable Kotlin will do the null check for you, and that way the reader can see the type as well - it's not obvious here from the surrounding context.

Reply:
Good point. Added explicit types across the codebase. I was spoiled by Intellij, as it shows the type automatically if not specified.